scripts/Inelastic/CrystalField/fitting.py

```diff
@@ -1444,13 +1502,67 @@ def _fit_multi(self):
         self.model.update(function)
         self.model.chi2 = alg.getProperty('OutputChi2overDoF').value
 
+    def fit_sp(self, Solver: str, Options: dict = None) -> None:
+        """
+        Run scipy.optimize.minimize algorithm for CEF parameters only. Update function parameters.
+        """
+        self.check_consistency()
+        if isinstance(self.model, CrystalFieldMultiSite):
+            fun = self.model.function
+        else:
+            fun = self._function
+        cef_fixed = []
+        for par_id in [id for id in range(fun.nParams()) if fun.isFixed(id)]:
+            parName = fun.getParamName(par_id)
+            if parName in CrystalField.field_parameter_names or "IntensityScaling" in parName:
+                cef_fixed.append(par_id)
+        x0 = []
+        fun.setAttributeValue('FixAllPeaks', True)
+        for par_id in self._free_cef_parameters:
+            x0.append(fun.getParameterValue(par_id))
+            fun.fixParameter(par_id)
+        opt = {'disp': False}
+        if Options is not None:
+            opt = Options
+        for (x, pos) in zip(x0, self._free_cef_parameters):
+            fun.removeTie(pos)
+            res = sp.minimize(self._evaluate_cf, [x], args=(fun, pos), method=Solver, options=opt)
+            fun.fixParameter(pos)
+            if res.success:
+                if self._function is not None:
+                    if isinstance(res.x, list):
+                        fun.setParameter(pos, float(res.x[0]))
+                    else:
+                        fun.setParameter(pos, float(res.x))
+        self.model.update(fun)
+
+    def _evaluate_cf(self, x0: float, fun: IFunction, cef_pos: int) -> float:
+        fun.setParameter(cef_pos, x0[0])
+        if isinstance(self._input_workspace, list):
+            ws_kwargs = {}
+            ws_kwargs['InputWorkspace'] = self._input_workspace[0]
+            i = 1
+            for workspace in self._input_workspace[1:]:
+                ws_kwargs['InputWorkspace_{}'.format(i)] = workspace
+                i += 1
+            # clean up multispectrum function to prevent problems during evaluation
+            # e.g. remove FWHMX0/FWHMY0 and FWHMX1/FWHMY1
+            fun_str = re.sub(r'FWHM[X|Y]\d+=\(\),', '', str(fun))
```

Comment:
I don't want to hold this up so it can be tested by scientists, especially as it's experimental, but I'd highly encourage any fixed regex strings to be compiled once using `re.compile` and stored at the module level - there is an example at the top of this module. Especially if we are evaluating the function then it would be good to avoid any extraneous work.
Might I also suggest capturing this in a single function so you could do something like
```python
return CalculateChiSquared(create_multispectrum_function_str(fun), **ws_kwargs)[1]
```